### PR TITLE
FT0 info table and FT0 AOD QA

### PR DESCRIPTION
**Pull request by AllaMaevskaya**

### CODE REVIEW

**Comment by iarsene** [comment]
Hi Ala,
In the QA task, the filling of histograms is commented out. Is this intended like that for now?

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisTask.h"
+#include "Common/DataModel/EventSelection.h"
+#include "Framework/AnalysisDataModel.h"
+#include "Common/CCDB/TriggerAliases.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "CommonConstants/LHCConstants.h"
+#include "Common/DataModel/T0infoTable.h"
+#include "iostream"
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:30]: Why is this needed?

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          t0AC = (t0A + t0C) / 2;
+        }
+        if (ora) {
+          t0A += vertex_corr;
+        }
+        if (orc) {
+          t0C -= vertex_corr;
+        }
+      }
+      LOGF(info, " T0 collision time = %f, T0A = %f, T0C = %f, T0 vertex = %f PV = %f", t0A, t0C, t0AC, vertexT0, vertexPV);
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:84]: please change to debug

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    //      LOGF(fatal, "CCDB host %s is n
+    ot reacheable, cannot go forward", url.value.data());
+    //    }
+  }
+  */
+  void process(BCsWithMatchings const& bcs, soa::Join<aod::Collisions, aod::EvSels> const& collisions, aod::FT0s const& ft0s)
+  {
+    //   auto caliboffsets = ccdb.get<o2::ft0::GlobalOffsetsCalibrationObject>("FT0/Calibration/GlobalOffsets");
+    LOG(INFO) << "infoTableFT0producer";
+    float c = 29.9792458; // cm/ns
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:56]: This constant is already defined in O2

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    //    if (!ccdb_api.isHostReachable()) {
+    //      LOGF(fatal, "CCDB host %s is n
+    ot reacheable, cannot go forward", url.value.data());
+    //    }
+  }
+  */
+  void process(BCsWithMatchings const& bcs, soa::Join<aod::Collisions, aod::EvSels> const& collisions, aod::FT0s const& ft0s)
+  {
+    //   auto caliboffsets = ccdb.get<o2::ft0::GlobalOffsetsCalibrationObject>("FT0/Calibration/GlobalOffsets");
+    LOG(INFO) << "infoTableFT0producer";
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:55]: Please remove this line

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      if (foundFT0 != -1) {
+        auto ft0 = ft0s.iteratorAt(foundFT0);
+        t0A = ft0.timeA();
+        t0C = ft0.timeC();
+        int triggersignals = ft0.triggerMask();
+        bool ora = (triggersignals & (1 << 0)) != 0;
+        bool orc = (triggersignals & (1 << 1)) != 0;
+        bool vertexTrigger = (triggersignals & (1 << 2)) != 0;
+        LOGF(info, "triggers OrA %i OrC %i vertexTrigger %i", ora, orc, vertexTrigger);
+        LOGF(info, " T0A = %f, T0C %f, vertex_corr %f, triggersignals %i", t0A, t0C, vertex_corr, triggersignals);
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:72]: also debug here

File: Common/TableProducer/infoTableFT0producer.cxx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        }
+      }
+      LOGF(info, " T0 collision time = %f, T0A = %f, T0C = %f, T0 vertex = %f PV = %f", t0A, t0C, t0AC, vertexT0, vertexPV);
+      table(t0A, t0C, t0AC, vertexT0);
+    }
+  }
+};
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  LOG(INFO) << "WorkflowSpec defineDataProcessin";
```
Comment (jgrosseo) [inline Common/TableProducer/infoTableFT0producer.cxx:91]: remove

File: Common/DataModel/T0infoTable.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+#define O2_T0_INFO_TABLE_H_
+
+#include "Framework/AnalysisDataModel.h"
+
+namespace o2::aod
+{
+namespace infoFT0
+{
+DECLARE_SOA_COLUMN(T0A, t0A, float);           //! Collision time A-side, corrected with primary vertex
+DECLARE_SOA_COLUMN(T0C, t0C, float);           //! Collision time A-side, corrected with primary vertex
```
Comment (jgrosseo) [inline Common/DataModel/T0infoTable.h:21]: C- side :)

File: Common/DataModel/T0infoTable.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+DECLARE_SOA_COLUMN(T0AC, t0AC, float);         //! Collision time (A+C)/2
```
[the remark this answers is not in the document]
Comment (AllaMaevskaya) [reply Common/DataModel/T0infoTable.h:22]: No, I calculate (T0A+T0C)/2  at line 57, then, afterwards, at  line 59
I add correction to one side and subtract from the opposite.
So, we got 3 collision times for TOF: 1. the best one (T0A+T0C)/2, 2. T0A, 3. T0C. All three are not depend on vertex position: 1. - naturally, 2&3  were corrected

[the remark this answers is not in the document]
Comment (jgrosseo) [reply Common/DataModel/T0infoTable.h:22]: Alla, it is the same isn't it?

AC=(A+C)/2
A'=A+V
C'=C-V
--> (A'+C')/2 = (A+V+C-V)/2 = (A+C)/2 = AC

What am I overlooking?

[the remark this answers is not in the document]
Comment (AllaMaevskaya) [reply Common/DataModel/T0infoTable.h:22]: A and C should be used separately. When both are exist - AC using.  A or  C are used in low multiplicity events when only one of them exist. A and C has worse resolution then AC.

[the remark this answers is not in the document]
Comment (jgrosseo) [reply Common/DataModel/T0infoTable.h:22]: But would one use `AC` if only `A` or `C` exist?
My point is that it is mathematically identical if `ora` and `orc` are `true` in your code.
Could you tell me me where is the mistake in my calculation? Otherwise I fail to understand your point.

File: Common/TableProducer/FT0Table.cxx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        t0A = ft0.timeA();
+        t0C = ft0.timeC();
+        int triggersignals = ft0.triggerMask();
+        bool ora = (triggersignals & (1 << 0)) != 0;
+        bool orc = (triggersignals & (1 << 1)) != 0;
+        bool vertexTrigger = (triggersignals & (1 << 2)) != 0;
+        LOGF(debug, "triggers OrA %i OrC %i vertexTrigger %i", ora, orc, vertexTrigger);
+        LOGF(debug, " T0A = %f, T0C %f, vertex_corr %f, triggersignals %i", t0A, t0C, vertex_corr, triggersignals);
+        if (vertexTrigger) {
+          vertexT0 = o2::constants::physics::LightSpeedCm2NS * (t0C - t0A) / 2;
```
[the remark this answers is not in the document]
Comment (jgrosseo) [reply Common/TableProducer/FT0Table.cxx:56]: See this example: https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/include/Framework/AnalysisDataModel.h#L104

File: Common/DataModel/T0infoTable.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (AllaMaevskaya) [reply Common/DataModel/T0infoTable.h:22]: AC is collision time with very good resolution. But in events with low multiplicity (due to acceptance) could be that only A or only C exits. But average time on each side depend on vertex, yes? So to use A (or C)   as collision time we need to remove vertex dependence. Resolution of these collision time would be much worse, but still usable. In Run2 for p-p AC with resolution ~50ps were exist in 50% events, only A (with resolution ~75ps) added ~10% events , only C (with resolution ~70ps) provided additional 15%. It was rather important for TOF.

[the remark this answers is not in the document]
Comment (jgrosseo) [reply Common/DataModel/T0infoTable.h:22]: Could you tell me me where is the mistake in my calculation? Otherwise I fail to understand your point.

File: Common/TableProducer/FT0Table.cxx (hunk 9 above)
[the remark this answers is not in the document]
Comment (AllaMaevskaya) [reply Common/TableProducer/FT0Table.cxx:56]: Will see, I think we will need

File: Common/DataModel/T0infoTable.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (AllaMaevskaya) [reply Common/DataModel/T0infoTable.h:22]: Yes, is it possible to use primary vertex in dynamic column? I think AC and vertex could be dynamic for sure.

File: Common/DataModel/T0infoTable.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace o2::aod
+{
+namespace infoFT0
+{
+DECLARE_SOA_COLUMN(T0A, t0A, float);
+//! Collision time A-side, corrected with primary vertex
+DECLARE_SOA_COLUMN(T0C, t0C, float);
+//! Collision time C-side, corrected with primary vertex
+DECLARE_SOA_COLUMN(T0vertex, t0vertex, float);
```
Comment (jgrosseo) [inline Common/DataModel/T0infoTable.h:27]: Thanks Alla!
Can you also put a dynamic column for t0vertex?
